src/Microsoft.DotNet.Darc/src/DarcLib/Models/Darc/DependencyGraph.cs

```diff
@@ -143,14 +190,181 @@ public class DependencyGraph
+                    if (newestBuildWithChannel != null)
+                    {
+                        int channelId = newestBuildWithChannel.Channels.First().Id.Value;
+                        // Just choose the first channel. This algorithm is mostly just heuristic.
+                        string latestCommitKey = $"{node.RepoUri}@{channelId}";
+                        string latestCommit = null;
+                        if (!latestCommitCache.TryGetValue(latestCommitKey, out latestCommit))
+                        {
+                            // Look up latest build in the channel
+                            var latestBuild = await barOnlyRemote.GetLatestBuildAsync(node.RepoUri, channelId);
+                            // Could be null, if the only build was removed from the channel
+                            if (latestBuild != null)
+                            {
+                                latestCommit = latestBuild.Commit;
+                            }
+                            // Add to cache
+                            latestCommitCache.Add(latestCommitKey, latestCommit);
+                        }
+
+                        // Perform diff if there is a latest commit.
+                        if (!string.IsNullOrEmpty(latestCommit))
+                        {
+                            IRemote repoRemote = remoteFactory.GetRemote(node.RepoUri, logger);
+                            // This will return a no-diff if latestCommit == node.Commit
+                            node.DiffFrom = await repoRemote.GitDiffAsync(node.RepoUri, latestCommit, node.Commit);
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        ///     Diff each node in the graph against the latest build in
+        ///     the graph.
+        /// </summary>
+        /// <param name="remoteFactory"></param>
+        /// <param name="logger"></param>
+        /// <param name="nodeCache"></param>
+        /// <param name="visitedRepoUriNodes"></param>
+        /// <returns></returns>
+        private static async Task DoLatestInGraphNodeDiffAsync(
+            IRemoteFactory remoteFactory,
+            ILogger logger,
+            Dictionary<string, DependencyGraphNode> nodeCache,
+            Dictionary<string, DependencyGraphNode> visitedRepoUriNodes)
+        {
+            logger.LogInformation("Running latest in graph node diff.");
+
+            // Find the build of each repo in the graph, then
+            // get the diff info from the latest
+            Dictionary<string, DependencyGraphNode>.KeyCollection uniqueRepos = visitedRepoUriNodes.Keys;
+            foreach (string repo in uniqueRepos)
+            {
+                // Get all nodes with this value
+                IEnumerable<DependencyGraphNode> nodes = nodeCache.Values.Where(n => n.RepoUri == repo);
+                // If only one, determine latest
+                if (nodes.Count() > 1)
```

Comment:
>nodes [](start = 20, length = 5)

Multiply enumerating IEnumerables isn't "safe". Sometimes it's ok, but in general, if you need to do multiple things with a set of things, it should be enumerated once to a concrete type and that used instead.  Usually that just means adding `.ToList()` to the end of LINQ to wrap it all up.

Reply:
What do you mean by not "safe"?